recipes_source/mobile_perf.rst

```diff
@@ -0,0 +1,191 @@
+
+Model preparation
+-----------------
+
+Next recipes you can take (offline) while preparing the model
+to have an optimized model that will probably have shorter execution time
+(higher performance, lower latency) on the mobile device.
+
+1. Use torch.utils.mobile_optimizer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Torch mobile_optimizer package does several optimizations with the model,
+which will help to conv2d and linear operations.
+It pre-packs model weights in an optimized format and fuses ops above with relu
+if it is the next operation.
+
+::
+
+  from torch.utils.mobile_optimizer import optimize_for_mobile
+  traced_model = torch.jit.load("input_model_path")
+  optimized_model = optimize_for_mobile(traced_model)
+  torch.jit.save(optimized_model, "output_model_path")
+
+2. Fuse operators using ``torch.quantization.fuse_modules``
+Do not be confused that fuse_modules is in the quantization package.
+It works for all types of torch script modules.
+``torch.quantization.fuse_modules`` fuses a list of modules into a single module.
+It fuses only the following sequence of modules:
+
+- Convolution, Batch normalization
+- Convolution, Batch normalization, Relu
+- Convolution, Relu
+- Linear, Relu
+
+This script shows how to fuse Batch Normalization in torchvision mobileNetV2.
+
+::
+
+  import torch
+  import torchvision
+
+  # ConvBNReLU: nn.Sequential[nn.Conv2d, nn.BatchNorm2d, nn.ReLU]
+  # InvertedResidual: nn.Sequential[ConvBnReLU, ConvBNReLU, nn.Conv2d, nn.BatchNorm2d]
+  def fuse_model(model):
+      for idx, m in enumerate(model.modules()):
+          name = m._get_name();
+          if name == 'ConvBNReLU':
+              torch.quantization.fuse_modules(m, ['0', '1'], inplace=True)
+          if name == 'InvertedResidual':
+              for idx in range(len(m.conv)):
+                  if type(m.conv[idx]) == torch.nn.Conv2d:
+                      torch.quantization.fuse_modules(m.conv, [str(idx), str(idx + 1)], inplace=True)
+
+  m = torchvision.models.mobilenet_v2(pretrained=True)
+  m.eval()
+  fuse_model(m)
+  torch.jit.trace(m, torch.rand(1, 3, 224, 224)).save("mobilenetV2-bnfused.pt")
+
+Quantization
+------------
+
+
+3. Try a quantized version of your model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can find more about PyTorch quantization in
+`the dedicated tutorial <https://pytorch.org/blog/introduction-to-quantization-on-pytorch/>`_.
+
+Quantization of the model not only moves computation to int8,
+but also reduces the size of your model on a disk.
+That size reduction helps to reduce disk read operations during the first load of the model and decreases the amount of RAM.
+Both of those resources can be crucial for the performance of mobile applications.
+
+::
+
+  import torch
+  import torchvision
+
+  supported_qengines = torch.backends.quantized.supported_engines
+  print(supported_qengines)
+  model = torchvision.models.quantization.__dict__['mobilenet_v2'](pretrained=True, quantize=True)
```

Comment:
Two things:

1) Is this the quantization workflow we want to show? This looks like it's just pulling down a pre-quantized version of the model, in which case this recipe is only useful if you want to quantize the torchvision models. In the general case, I'd think people would want to be able to quantize *their own* trained models for mobile deployment.
2) On MacOS, this line throws a warning:

```
/Users/bradheintz/anaconda2/envs/pyto16pre/lib/python3.8/site-packages/torch/nn/quantized/modules/utils.py:8: UserWarning: 0quantize_tensor_per_tensor_affine current rounding mode is not set to round-to-nearest-ties-to-even (FE_TONEAREST). This will cause accuracy issues in quantized models. (Triggered internally at  ../aten/src/ATen/native/quantized/affine_quantizer.cpp:25.)
  qweight = torch.quantize_per_tensor(
/Users/bradheintz/anaconda2/envs/pyto16pre/lib/python3.8/site-packages/torch/quantization/observer.py:134: UserWarning: must run observer before calling calculate_qparams.                                    Returning default scale and zero point
  warnings.warn(
```

(And yes, it really looks like that in my terminal.)

Reply:
We should show a workflow where we start with a floating point model and then do the quantization. The steps are:
# Start with a fully trained floating point model
# The model code is modified to enable eager mode quantization, for more details
# please see the quantization tutorials at: https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html

model = torchvision.models.quantization.__dict__['resnet18'](pretrained=True, quantize=False)
torch.backends.quantized.engine='qnnpack'
# We convert the float model with the appropriate_Qconfig
model.eval()
model.qconfig = torch.quantization.get_default_qconfig('qnnpack')
torch.quantization.prepare(model)
# Run model with representative data for calibration
# model(calibration_data)
torch.quantization.convert(model)
script_model = torch.jit.script(model)

# Export to mobile
script_model._save_for_lite_interpreter("model.bc")